Use correct column order for composite foreign keys

AndreasA opened the pull request:
Fixes wrong foreign key order. For now just an integration test.

|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | #4971

Commit:
Use correct column order for composite foreign keys

morozov commented on src/Platforms/MySQLPlatform.php:
```diff
@@ -185,7 +185,7 @@ public function getListTableForeignKeysSQL($table, $database = null)
 
         $sql = 'SELECT DISTINCT k.`CONSTRAINT_NAME`, k.`COLUMN_NAME`, k.`REFERENCED_TABLE_NAME`, ' .
-               'k.`REFERENCED_COLUMN_NAME` /*!50116 , c.update_rule, c.delete_rule */ ' .
+               'k.`REFERENCED_COLUMN_NAME`, k.`ORDINAL_POSITION` /*!50116 , c.update_rule, c.delete_rule */ ' .
```
Do we need to select this column? Its value seems to be used only for ordering.

AndreasA replied:
@morozov Thanks for giving feedback. I will try to remove it again. But I think I was not able to order by that column without selecting it. But I will re-check tomorrow.

AndreasA replied:
@morozov I just re-checked it is required due to the DISTINCT in the select clause, otherwise I get the following error.
`3065 Expression #1 of ORDER BY clause is not in SELECT list, references column 'information_schema.k.ORDINAL_POSITION' which is not in SELECT list; this is incompatible with DISTINCT` which is because MySQL 8.0 uses `ONLY_FULL_GROUP_BY` by default.

morozov replied:
Oh, this is because of the `DISTINCT`. It's often a smell of something missing from the query but we can leave it as is for now. Thanks for the details.

AndreasA replied:
Not 100% sure the DISTINCT is actually necessary but I didn't want to touch that part as I would first need to understand why it has been put there in the first place.

morozov commented on src/Platforms/MySQLPlatform.php:
```diff
@@ -195,7 +195,8 @@ public function getListTableForeignKeysSQL($table, $database = null)
             . ' /*!50116 AND c.constraint_schema = ' . $databaseNameSql . ' */'
-            . ' AND k.`REFERENCED_COLUMN_NAME` is not NULL';
+            . ' AND k.`REFERENCED_COLUMN_NAME` is not NULL'
+            . ' ORDER BY k.`ORDINAL_POSITION` ASC';
```
`ASC` is redundant.

AndreasA replied:
@morozov I just checked other similar statements in MySQLPlatform and all of them add `ASC` at the end, see: https://github.com/doctrine/dbal/blob/3.1.3/src/Platforms/MySQLPlatform.php#L364
https://github.com/doctrine/dbal/blob/3.1.3/src/Platforms/MySQLPlatform.php#L157

personally, I also prefer to add the direction and not count on the default direction staying ASC, though it is extremely unlikely that MySQL will change that 😄 

However, as mentioned for other similar statements it is with ASC, so I would prefer to keep it, so it is in sync with the others.

morozov replied:
It is still redundant. The fact that this practice is used in other places doesn't justify using it here. Please remove it.

AndreasA replied:
I have now removed it.

Though, personally I think staying in line with existing code syntax used (in this case SQL for MySQL platform) should be more important here, than removing redundancy but I am fine with removing it. Also I wouldn't say it is redundant, it is specific. Redundant would be if the same ORDER BY would be added twice, after all there is the extermely unlikely scenario, that MySQL adds an option to change the default order by direction and somebody decides to use it but as mentioned extremely unlikely :smile:

morozov commented on tests/Functional/Schema/MySQLSchemaManagerTest.php:
```diff
@@ -544,6 +544,44 @@ public function testEnsureTableOptionsAreReflectedInMetadata(): void
+        $this->connection->executeStatement('DROP TABLE IF EXISTS test_table_metadata');
+
+        $sql = <<<'SQL'
+CREATE TABLE test_table_foreign_key_referenced (
```
Please create the table using the `Schema` API and move the test to the base class. This scenario is valid for all supported platforms.

UPD: probably, except for SQLite. Use `if ($platform->supportsForeignKeyConstraints())` to skip, if necessary.

AndreasA replied:
@morozov So I should create it like I do with the `test_table_with_foreign_key`  table?

morozov replied:
Yes.

morozov commented on tests/Functional/Schema/MySQLSchemaManagerTest.php:
```diff
@@ -544,6 +544,44 @@ public function testEnsureTableOptionsAreReflectedInMetadata(): void
+
+        $tableFetched = $this->schemaManager->listTableDetails('test_table_with_foreign_key');
+
+        $diff = (new Comparator())->diffTable($tableFetched, $table);
```
Since the issue is not about the comparator, it's redundant in the test as well. It's sufficient to get the foreign key definition of the introspected table and assert that the columns are in the right order.

AndreasA replied:
@morozov Makes sense.

morozov reviewed (commented):
Thanks for the patch, @AndreasA. Please see a few comments inline.

AndreasA commented:
@morozov The threads should now be fixed or if not I added a comment why not.

I have now also manually executed the two create table statements on another machine with MySQL 8.0 and the retrieve foreign key SQL for it and the result was the same regarding the foreign key columns order. without ORDINAL_POSITION it was different there as well, so I think it is not just an issue - due to some configuration - for my MySQL installation. In any case using ORDINAL_POSITION should always work.

I have also checked and `ORDINAL_POSITION` was in that table since 5.6 - e.g. in case of MariaDB https://dev.mysql.com/doc/refman/5.6/en/information-schema-key-column-usage-table.html

By now moving the test to be used for all platforms, I cannot be sure 100% if the issue is fixed for all of them. I have fixed it for the SQLSrv tests now because that workflow was automatically started. 

I just hope that all others are already retrieving the data in the correct order 😄  From a short glance they should do so but the tests should make sure.

One other thing: I have created the PR for 3.x but should I actually create a PR for 2.x which can then be merged into 3.x? though I doubt it is an issue that occurs that often, otherwise I think it would have been addressed earlier.

AndreasA commented on src/Platforms/SQLServer2012Platform.php:
```diff
@@ -1010,7 +1010,8 @@ public function getListTableForeignKeysSQL($table, $database = null)
                 WHERE ' .
-                $this->getTableWhereClause($table, 'SCHEMA_NAME (f.schema_id)', 'OBJECT_NAME (f.parent_object_id)');
+                $this->getTableWhereClause($table, 'SCHEMA_NAME (f.schema_id)', 'OBJECT_NAME (f.parent_object_id)') .
+                ' ORDER BY fc.constraint_column_id';
```
in SQLSrv some order bys add `ASC` and some don't, so I have not added it here to be in sync with your initial suggestion regarding MySQL. However, in the MySQL platform ASC is always there so I think we should add it there.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -1418,6 +1419,58 @@ public function testCommentInTable(): void
+        $table->addOption('comment', 'Test foreign key referenced table');
+
+        $table->addColumn('col1', Types::INTEGER, ['notNull' => true]);
+        $table->addColumn('col2', Types::INTEGER, ['notNull' => true]);
```
Are the `notNull` options required for the test? If not, please remove them.

AndreasA replied:
Are removed now.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -1418,6 +1419,58 @@ public function testCommentInTable(): void
+
+        $table->addOption('collate', 'utf8_general_ci');
+        $table->addOption('charset', 'utf8');
+        $table->addOption('comment', 'Test foreign key referenced table');
```
All the options above are irrelevant to the test. Please remove them.

AndreasA replied:
Are removed now.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -1418,6 +1419,58 @@ public function testCommentInTable(): void
+        $table->addColumn('col2', Types::INTEGER, ['notNull' => true]);
+        $table->setPrimaryKey($foreignColumns);
+
+        $this->schemaManager->dropAndCreateTable($table);
```
I think `dropAndCreate()` won't work well for the tables connected by a foreign key. Please try running this test as follows  and make sure that both runs pass:
```shell
phpunit \
  --repeat 2 \
  --filter testCreatedCompositeForeignKeyOrderIsCorrectAfterCreation \
  tests/Functional/Schema/MySQLSchemaManagerTest.php
```

You may need to use something like this instead: https://github.com/doctrine/dbal/blob/81c54e8caf8b959609fa6dab244904c54f0181fc/tests/Functional/Schema/OracleSchemaManagerTest.php#L128-L132

For that, you'll need to temporarily comment out this line which is another PITA:
https://github.com/doctrine/dbal/blob/81c54e8caf8b959609fa6dab244904c54f0181fc/tests/Functional/Schema/MySQLSchemaManagerTest.php#L21-L21

AndreasA replied:
@morozov Thanks for the hint. However, dropping did not help. I had to first remove the foreign key, otherwise it did not work. Probably could have kept the dropAndCreate now but I have kept them separate now.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -1418,6 +1419,58 @@ public function testCommentInTable(): void
+        self::assertSame(
+            $localColumns,
+            $table->getForeignKey('fk_test_foreign_key')->getLocalColumns(),
+            'Order for the foreign key local columns should match the specified order for the foreign key constraint.'
```
Personally, I find custom assertion messages doing more harm than good. An assertion may fail not only because the order doesn't match, so this message will be misleading. I'd remove it completely.

AndreasA replied:
Done. Though phpunit would output the differences too but I am happy with either solution. Personally, I very seldom use custom error messages too, but I saw some other tests using  them, so I thought it might be expected.

morozov commented:
> I just hope that all others are already retrieving the data in the correct order 😄 From a short glance they should do so but the tests should make sure.

I approved the test run.

> One other thing: I have created the PR for 3.x but should I actually create a PR for 2.x which can then be merged into 3.x? though I doubt it is an issue that occurs that often, otherwise I think it would have been addressed earlier.

There's no need to backport the fix to `2.x` unless you really need to.

AndreasA commented:
@morozov I have now done the necessary adjustments. Can you re-check? Not sure if the test run is running against the lastest verison.

morozov commented:
You'll need to disable the test on SQLite (see https://github.com/doctrine/dbal/pull/4978#discussion_r746828703). Its failure prevents other jobs from running.

AndreasA commented:
> You'll need to disable the test on SQLite (see [#4978 (comment)](https://github.com/doctrine/dbal/pull/4978#discussion_r746828703)). Its failure prevents other jobs from running.

Ah yes. Missed that part and have added it now.

AndreasA commented:
@morozov 

> You'll need to disable the test on SQLite (see [#4978 (comment)](https://github.com/doctrine/dbal/pull/4978#discussion_r746828703)). Its failure prevents other jobs from running.

Ah yes. Missed that part and have added it now. Thanks for letting me know.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -1418,6 +1419,53 @@ public function testCommentInTable(): void
+        $localColumns              = ['child_col2', 'child_col1'];
+        $foreignColumns            = ['col2', 'col1'];
+
+        $this->schemaManager->tryMethod('dropForeignKey', $foreignKey, $tableWithForeignKey);
```
Dropping the FK shouldn't be necessary. Instead, the order of `DROP TABLE`'s should be changed: first the referencing, then the referenced.

AndreasA replied:
@morozov Ah, damn, of course it shouldn't matter as long as the order in which they are dropped is correct. 

I guess it is indeed late 😄 

Fixed this now and also shortened the names so the OCI tests are hopefully successful as well.

AndreasA replied:
btw. any reason why this line ` Type::addType('point', PointType::class); ` isn't something like:
`Type::hasType('point') ? Type::overrideType('point', PointType::class) : Type::addType('point', PointType::class);`

morozov replied:
It probably should be like that. Nobody bothered to fix it yet.

derrabus commented:
Can you have a look at the test failures on Oracle and DB2?

AndreasA commented:
@derrabus I just did, apparantly the column names are always uppercase there but I specify them originally lowercase. I guess the easies solution would be to create them uppercase for now as I don't really care about the case just the order.

Or I ignore the case during the assertion. Not sure which way to go

morozov commented:
Most tests create the columns in the lower case and use `strtolower()` on the introspection results before comparison.

AndreasA commented:
@morozov @derrabus Added a corresponding strtolower conversion for the result comparison now. Hopefully, the tests will work now.

morozov commented:
@AndreasA, the patch looks good to me. Please squash all commits, and we can merge it.

AndreasA commented:
@morozov Done.

morozov commented:
Thanks, @AndreasA!
